tools/Common.psm1

```diff
@@ -0,0 +1,553 @@
+        try
+        {
+            $expressionResult = Invoke-Expression $CheckExpression
+            $stringOutput = $expressionResult
+            $loop = $expressionResult
+            if ($Negate)
+            {
+                $loop = !$expressionResult
+            }
+        }
+        catch
+        {
+            Write-Verbose -Message "ERROR running command: ${CheckExpression}" -Verbose
+            Write-Verbose -Message $_.ScriptStackTrace -Verbose
+            Write-Verbose -Message "$($_ | Out-String)" -Verbose
+            $loop = $true
+        }
+
+        $invalidToString = $false
+        try
+        {
+            $stringOutput = $expressionResult.toString()
+        }
+        catch
+        {
+            $invalidToString = $true
+            $stringOutput = $expressionResult | Out-String
+            Write-Verbose -Message "'${CheckExpression}' does not have a valid toString(), and has an Out-String of '${stringOutput}'." -Verbose
+        }
+
+        if (!$invalidToString)
+        {
+            Write-Verbose -Message "'${CheckExpression}' evaluated to '${stringOutput}'." -Verbose
+        }
+
+        # $loop is $true when $CheckExpression evaluates to truthy value, but if $Negate is true, then that logic is reversed
+        # $loop is also $true if $CheckExpression evaluation throws an exception.
+        if ($loop)
+        {
+            Write-Verbose -Message "Waited for ${currentSeconds} seconds. Waiting ${SleepSecondsUnit} more seconds and then running again..." -Verbose
+            $currentSeconds += $SleepSecondsUnit
+            Start-Sleep -Seconds $SleepSecondsUnit
+        }
+        else
+        {
+            break
+        }
+    }
+
+    if ($currentSeconds -ge ($MaxSleepSeconds + 1))
+    {
+        Write-Verbose -Message "'${CheckExpression}' still evaluates to '${stringOutput}' after timeout of ${MaxSleepSeconds} seconds." -Verbose
+    }
+    else
+    {
+        Write-Verbose -Message "'${CheckExpression}' evaluated to a value of '${stringOutput}' before timeout of ${MaxSleepSeconds} seconds." -Verbose
+    }
+
+    $expressionResult
+}
+
+# Creates a new service principal client Id and secret. Only required if a service principal does not already exist. Requires admin privileges to create.
+function New-AzureAppSp
+{
+    [CmdletBinding(DefaultParameterSetName = "notAdfs")]
+    param
+    (
+        [Parameter(ParameterSetName = "Aad", Mandatory = $true)]
+        [ValidateNotNullOrEmpty()]
+        [string] $ServicePrincipalName,
+        [Parameter(ParameterSetName = "Adfs", Mandatory = $true)]
+        [ValidateNotNullOrEmpty()]
+        [string] $ApplicationName,
+        [Parameter(ParameterSetName = "Aad", Mandatory = $false)]
+        [Parameter(ParameterSetName = "Adfs", Mandatory = $false)]
+        [ValidateNotNullOrEmpty()]
+        [string] $ErcsVmDnsName= "AzS-ERCS01",
```

Comment:
can we figure out host name on 4 node env, so this will work in both single node and 4 node?

Reply:
The VMs are numbered from 1 to 3 or so. So the first one should always exist.